### PR TITLE
Fix wrong compiler path on windows when using andrdoid ndk and MesonToolchain

### PR DESCRIPTION
**Pull request by elvisdukaj**
Changelog: Bugfix: Fix cross-compilation to Android from Windows when using ``MesonToolchain``.
Docs: Omit

- [x] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

# Description 

Trying to build `lcms/2.14` on Windows using Android NDK results in a build error.

```
..\src\meson.build:1:0: ERROR: Failed running 'C:\\Users\\edukaj\\.conan2\\p\\b\\andro0a7659df7e598\\p\\bin\\toolchains\\llvm\\prebuilt\\windows-x86_64\\bin\\aarch64-linux-android29-clang', binary or interpreter not executable.
Possibly wrong architecture or the executable bit is not set.
```

The issue is that on Windows the right path for the toolchain should end with `.cmd`. 

I opened an issue on [conan-center-index](https://github.com/conan-io/conan-center-index/issues/22909), and I am proposing a fix here.

# Note

I accidentaly close https://github.com/conan-io/conan/pull/15760 and I had to create a new one here. I am sorry for the inconvinience!

### CODE REVIEW

File: conan/tools/meson/toolchain.py [hunk 1]
```diff
@@ -329,8 +329,10 @@ def _resolve_android_cross_compilation(self):
                           'armv8': 'aarch64-linux-android',
                           'x86': 'i686-linux-android',
                           'x86_64': 'x86_64-linux-android'}.get(arch)
-        self.c = os.path.join(ndk_bin, "{}{}-clang".format(android_target, android_api_level))
-        self.cpp = os.path.join(ndk_bin, "{}{}-clang++".format(android_target, android_api_level))
+        os_build = self._conanfile.settings_build.get_safe('os')
+        compiler_extension = ".cmd" if os_build == "Windows" else ""
```
Comment (memsharded) [inline conan/tools/meson/toolchain.py:339]: I am trying this in my local machine, and it is failing. 
It happens that ``os_build`` is ``windows`` in lowercase, and it doesn't match.

File: conans/test/functional/toolchains/meson/test_cross_compilation.py [hunk 2]
```diff
@@ -204,3 +204,76 @@ def test_android_meson_toolchain_cross_compiling(arch, expected_arch):
     if platform.system() == "Darwin":
         client.run_command('objdump -f "%s"' % libhello)
         assert "architecture: %s" % expected_arch in client.out
+
+
+@pytest.mark.parametrize("arch, expected_arch", [('armv8', 'aarch64'),
+                                                 ('armv7', 'arm'),
+                                                 ('x86', 'i386'),
+                                                 ('x86_64', 'x86_64')])
```
Comment (memsharded) [inline conans/test/functional/toolchains/meson/test_cross_compilation.py]: No need to do the full test of architectures, this is slow, takes too much time on CI.
Using just 1 is enough to check that the ``.cmd`` thing works fine in Windows

**Review by memsharded — COMMENTED** [review]
Thanks very much for your contribution

File: conans/test/functional/toolchains/meson/test_cross_compilation.py (hunk 2 above)
Comment (elvisdukaj) [reply conans/test/functional/toolchains/meson/test_cross_compilation.py]: I managed to have green tests locally! I will submit the fix soon :). I am testing only my specific test case. It was pretty tricky to get things right!

Comment (elvisdukaj) [reply conans/test/functional/toolchains/meson/test_cross_compilation.py]: > No need to do the full test of architectures, this is slow, takes too much time on CI.
Using just 1 is enough to check that the .cmd thing works fine in Windows

I am writing tests in a more generic way than the existing one with the hope that we could have only one test case for the cross-platform.

**Comment by elvisdukaj** [comment]
Ok I've no idea why is failing again... I can reduce the scope of the tests to windows only, for armv8... The point is that toolchain path can be very different based on the arch, but maybe this is tested somewhere else.

**Comment by memsharded** [comment]
> Ok I've no idea why is failing again... I can reduce the scope of the tests to windows only, for armv8... The point is that toolchain path can be very different based on the arch, but maybe this is tested somewhere else.

I think we don't have AndroidNDK in the Windows CI agent, @czoido?

**Comment by elvisdukaj** [comment]
It would be nice to have it! Since I started to test conan on windows with android ndk, I’m hitting so many issues… I’ve created so many PR in the last days! It seems not a common thing

**Review by memsharded — COMMENTED** [review]
As we don't have Android in Windows at the moment, I was suggesting:

- Have the test exclusively for OSX-Darwin, as this is the one we have now in CI
- We can run the test locally on Linux and Windows to manually validate
- We can move the PR forward if it passes CI with OSX and we test it locally too

I'll discuss with the team the possibility to have the AndroidNDK at least on Windows or Linux to improve the coverage. I think Windows might be easier.

**Comment by memsharded** [comment]
I am doing some simplifications in https://github.com/conan-io/conan/pull/15808 in order to prepare for having Android testing in other platforms

**Comment by elvisdukaj** [comment]
Should I comment the tests on Windows for now?

**Comment by memsharded** [comment]
Not commenting out, but enabling it only for Darwin/OSX. That should pass, we already have other tests using AndroidNDK in OSX without issues.

Then, we can just comment out the skipif line in the code and run it locally to verify.

I have tried to contribute to your PR, but you have contributions from maintainers blocked apparently.

**Comment by elvisdukaj** [comment]
Ok interesting, you should be able to contribute:

image

> Not commenting out, but enabling it only for Darwin/OSX. That should pass, we already have other tests using AndroidNDK in OSX without issues.
> Then, we can just comment out the skipif line in the code and run it locally to verify.

Okay I'll do as you suggested :)

File: conan/tools/meson/toolchain.py (hunk 1 above)
Comment (elvisdukaj) [reply conan/tools/meson/toolchain.py:339]: I removed this part, so this thread is not applicable anymore.

**Comment by elvisdukaj** [comment]
I also tested with MacOS on intel, with meson and android-ndk installed from homebrew end tests are green. 

```Python
tools_locations = {
    'meson': {
        "disabled": False,
        "platform": "Darwin",
        "default": "1.3.2",
        "1.3.2": {
            'path': {
                "Darwin": "/usr/local/bin"
            }
        }
    },
    'android_ndk': {
        "disabled": False,
        "platform": "Darwin",
        "default": "26c",
        "26c": {
            "path": {
                "Darwin": "/usr/local/share/android-ndk"
            }
        },
        "exe": "ndk-build",
    }
}

tools_environments = {
    'android_ndk': {'Darwin': {'TEST_CONAN_ANDROID_NDK': '/usr/local/share/android-ndk'}}
}
```

I cannot explain why the CI is still red, can you share the log?

**Comment by memsharded** [comment]
> Ok interesting, you should be able to contribute:

Ok, the reason was that you named the repo ``conan-io`` instead of ``conan``, and that was failing

**Comment by elvisdukaj** [comment]
So that means I need to close this MR and create a new one? I renamed the branch, let's hope is enough 🙈🙈🙈

**Comment by memsharded** [comment]
> So that means I need to close this MR and create a new one? I renamed the branch, let's hope is enough 🙈🙈🙈

No, no no need to do that. It was just that I added the remote with a wrong name. Changing the remote name to ``conan-io`` works

**Review by memsharded — APPROVED** [review]
The fix is correct,  I have validated it in Windows.

The tests might still fail on CI but that is on our side, so don't worry about it, I'll take care it to be merged for next 2.2. 

Thanks for your contribution!

**Comment by elvisdukaj** [comment]
@memsharded great news! Looking forward to get rid off my fix branch and use the upstream version. I have also some projects that is using conan v1, I will need to check if the fix should be back ported as well in my case.

**Comment by alexandermarinissen** [comment]
Hi! I'm experiencing the same issue, but in Conan 1.65.0. Has the fix been backported? I don't see the changes when I look at conan/tools/meson/toolchain.py in my local site-packages.

**Comment by memsharded** [comment]
Hi @alexandermarinissen 

Thanks for your question.

No, this fix has not been backported, this kind of thing is not being backported anymore to Conan 1.X by default by the maintainers. We have accepted some contributions to backport some, but also as an exception (reviewing, releasing, etc also takes time and resources, and Conan 2.0 is now more than 18 months old).

**Comment by alexandermarinissen** [comment]
Hi @memsharded, luckily the fix is quite small and I was able to apply it locally to my installation, so I'm no longer stuck :) Thanks for the quick reply!

**Comment by memsharded** [comment]
Sounds good, thanks for the feedback @alexandermarinissen 
Please just recall to try to prioritize moving to Conan 2, for example ConanCenter packages will stop getting updates for Conan 1 in the following months.

**Pull request merged**